hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/FlushTableProcedure.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.hadoop.hbase.procedure2.ProcedureStateSerializer;
+import org.apache.hadoop.hbase.procedure2.ProcedureSuspendedException;
+import org.apache.hadoop.hbase.procedure2.ProcedureYieldException;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.protobuf.UnsafeByteOperations;
+
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.FlushTableProcedureStateData;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.FlushTableState;
+
+@InterfaceAudience.Private
+public class FlushTableProcedure extends AbstractStateMachineTableProcedure<FlushTableState> {
+  private static final Logger LOG = LoggerFactory.getLogger(FlushTableProcedure.class);
+
+  private TableName tableName;
+
+  private byte[] columnFamily;
+
+  public FlushTableProcedure() {
+    super();
+  }
+
+  public FlushTableProcedure(MasterProcedureEnv env, TableName tableName) {
+    this(env, tableName, null);
+  }
+
+  public FlushTableProcedure(MasterProcedureEnv env, TableName tableName, byte[] columnFamily) {
+    super(env);
+    this.tableName = tableName;
+    this.columnFamily = columnFamily;
+  }
+
+  @Override
+  protected LockState acquireLock(MasterProcedureEnv env) {
+    // Here we don't acquire table lock because the flush operation and other operations (like
+    // split or merge) are not mutually exclusive. Region will flush memstore when being closed.
+    // It's safe even if we don't have lock. However, currently we are limited by the scheduling
+    // mechanism of the procedure scheduler and have to acquire table shared lock here. See
+    // HBASE-27905 for details.
+    if (env.getProcedureScheduler().waitTableSharedLock(this, getTableName())) {
+      return LockState.LOCK_EVENT_WAIT;
+    }
+    return LockState.LOCK_ACQUIRED;
+  }
+
+  @Override
+  protected void releaseLock(MasterProcedureEnv env) {
+    env.getProcedureScheduler().wakeTableSharedLock(this, getTableName());
+  }
+
+  @Override
+  protected Flow executeFromState(MasterProcedureEnv env, FlushTableState state)
+    throws ProcedureSuspendedException, ProcedureYieldException, InterruptedException {
+    LOG.info("{} execute state={}", this, state);
+
+    try {
+      switch (state) {
+        case FLUSH_TABLE_PREPARE:
+          preflightChecks(env, true);
+          setNextState(FlushTableState.FLUSH_TABLE_FLUSH_REGIONS);
+          return Flow.HAS_MORE_STATE;
+        case FLUSH_TABLE_FLUSH_REGIONS:
+          addChildProcedure(createFlushRegionProcedures(env));
+          return Flow.NO_MORE_STATE;
+        default:
+          throw new UnsupportedOperationException("unhandled state=" + state);
+      }
+    } catch (Exception e) {
+      setFailure("master-flush-table", e);
```

Comment:
Thank you for pointing this out Duo, it is necessary that we adjust it after we make the FlushTableProcedure not support rollback.

My original thought was this, when we are in the state **FLUSH_TABLE_PREPARE**, we need to confirm that the state of the table is **Online**, which may require an RPC to access the meta table, because the **TableStateManager** loads the state of the table lazily. if the request fails, an exception nay be thrown, so I originally thought that if the request fails, then just roll back the procedure. 

Since we do not support rollback now, let me fix this.